src/core/ddsc/tests/psmxif.c

```diff
@@ -0,0 +1,441 @@
+        CU_ASSERT_FATAL(dds_qget_psmx_instances(qos, &strs_len, &strs));
+        CU_ASSERT_FATAL(strs_len == 1 && strcmp(strs[0], "dummy_psmx") == 0);
+        free_strings(strs_len, strs);
+        dds_delete_qos(qos);
+      }
+      {
+        dds_entity* x = NULL;
+        dds_return_t rc = dds_entity_pin(reader1, &x);
+        CU_ASSERT_FATAL(rc == DDS_RETCODE_OK && dds_entity_kind(x) == DDS_KIND_READER);
+        struct dds_psmx_endpoints_set* endpt_set = &((dds_reader*)x)->m_endpoint.psmx_endpoints;
+        CU_ASSERT_FATAL(endpt_set->length == 1 && endpt_set->endpoints[0] == psmx_endpt_expected);
+        dds_entity_unpin(x);
+      }
+
+      create_unique_topic_name("shared_memory", topicname, sizeof(topicname));
+      psmx_topic_expected = (dds_psmx_topic_t*)dmock->topics._buffer + dmock->topics._length;
+      dds_entity_t topic2 = dds_create_topic(participant, &PsmxType1_desc, topicname, NULL, NULL);
+      CU_ASSERT_FATAL(topic2 > 0);
+      {
+        dds_entity* x = NULL;
+        dds_return_t rc = dds_entity_pin(topic2, &x);
+        CU_ASSERT_FATAL(rc == DDS_RETCODE_OK && dds_entity_kind(x) == DDS_KIND_TOPIC);
+        struct dds_psmx_topics_set* topics_set = &((dds_topic*)x)->m_ktopic->psmx_topics;
+        CU_ASSERT_FATAL(topics_set->length == 1 && topics_set->topics[0] == psmx_topic_expected);
+        dds_entity_unpin(x);
+      }
+
+      psmx_endpt_expected = (dds_psmx_endpoint_t*)dmock->endpoints._buffer + dmock->endpoints._length;
+      writer2 = dds_create_writer(participant, topic2, NULL, NULL);
+      CU_ASSERT_FATAL(writer2 > 0);
+      {
+        dds_entity* x = NULL;
+        dds_return_t rc = dds_entity_pin(writer2, &x);
+        CU_ASSERT_FATAL(rc == DDS_RETCODE_OK && dds_entity_kind(x) == DDS_KIND_WRITER);
+        struct dds_psmx_endpoints_set* endpt_set = &((dds_writer*)x)->m_endpoint.psmx_endpoints;
+        CU_ASSERT_FATAL(endpt_set->length == 1 && endpt_set->endpoints[0] == psmx_endpt_expected);
+        dds_entity_unpin(x);
+      }
+
+      psmx_endpt_expected = (dds_psmx_endpoint_t*)dmock->endpoints._buffer + dmock->endpoints._length;
+      reader2 = dds_create_reader(participant, topic2, NULL, NULL);
+      CU_ASSERT_FATAL(reader2 > 0);
+      {
+        dds_entity* x = NULL;
+        dds_return_t rc = dds_entity_pin(reader2, &x);
+        CU_ASSERT_FATAL(rc == DDS_RETCODE_OK && dds_entity_kind(x) == DDS_KIND_READER);
+        struct dds_psmx_endpoints_set* endpt_set = &((dds_reader*)x)->m_endpoint.psmx_endpoints;
+        CU_ASSERT_FATAL(endpt_set->length == 1 && endpt_set->endpoints[0] == psmx_endpt_expected);
+        dds_entity_unpin(x);
+      }
+    }
+
+    {
+      // Check that shared memory is enabled when it should, and not enabled when it shouldn't.
+      bool psmx_enabled;
+      psmx_enabled = endpoint_has_psmx_enabled(writer1);
+      CU_ASSERT_FATAL(psmx_enabled);
+      CU_ASSERT_FATAL(dds_is_shared_memory_available(writer1) == supports_shared_memory_expected);
```

Comment:
`supports_shared_memory_expected` is semantically a variable, but it is assigned only once, at declaration and its address is never taken. So it is a constant. I would prefer `CU_ASSERT_FATAL(dds_is_shared_memory_available(writer1))` in such a case.

However, here, you're trying to test `dds_is_shared_memory_available` but with the current set-up, an implementation of `dds_is_shared_memory_available` that simply returns `true` will pass. It is therefore not a good test.

A much better way would be to:
* Create a reader/writer that does not involve PSMX, verify that it returns false. (This would best be done in a different test.)
* Run this test once with the plugin claiming to support shared memory and verifying it returns true. (This is the current case.)
* Run this test once with the plugin claiming to *not* support shared memory and verifying it returns false. (This case is quite obviously missing.)

With the "direct" access to the dummy plugin, it is trivial to make the plugin claim one or the other.